diff --git a/src/cascadia/TerminalApp/AppActionHandlers.cpp b/src/cascadia/TerminalApp/AppActionHandlers.cpp
index 66372a3e9..d5157503b 100644
--- a/src/cascadia/TerminalApp/AppActionHandlers.cpp
+++ b/src/cascadia/TerminalApp/AppActionHandlers.cpp
@@ -722,21 +722,34 @@ namespace winrt::TerminalApp::implementation
                                         const ActionEventArgs& args)
     {
         std::optional<winrt::hstring> title;
+        winrt::com_ptr<Tab> targetTab{ nullptr };
 
         if (const auto& realArgs = args.ActionArgs().try_as<RenameTabArgs>())
         {
             title = realArgs.Title();
+            if (const auto& idx = realArgs.Index())
+            {
+                const auto tabIdx = idx.Value();
+                if (tabIdx < _tabs.Size())
+                {
+                    targetTab = _GetTabImpl(_tabs.GetAt(tabIdx));
+                }
+            }
+        }
+        if (!targetTab)
+        {
+            targetTab = _senderOrFocusedTab(sender);
         }
 
-        if (const auto activeTab{ _senderOrFocusedTab(sender) })
+        if (targetTab)
         {
             if (title.has_value())
             {
-                activeTab->SetTabText(title.value());
+                targetTab->SetTabText(title.value());
             }
             else
             {
-                activeTab->ResetTabText();
+                targetTab->ResetTabText();
             }
         }
         args.Handled(true);
diff --git a/src/cascadia/TerminalApp/AppCommandlineArgs.cpp b/src/cascadia/TerminalApp/AppCommandlineArgs.cpp
index c41eea543..6f5a17338 100644
--- a/src/cascadia/TerminalApp/AppCommandlineArgs.cpp
+++ b/src/cascadia/TerminalApp/AppCommandlineArgs.cpp
@@ -209,6 +209,8 @@ void AppCommandlineArgs::_buildParser()
     _buildSwapPaneParser();
     _buildFocusPaneParser();
     _buildSaveSnippetParser();
+    _buildCloseTabParser();
+    _buildRenameTabParser();
 }
 
 // Method Description:
@@ -766,6 +768,10 @@ bool AppCommandlineArgs::_noCommandsProvided()
 {
     return !(*_newTabCommand.subcommand ||
              *_newTabShort.subcommand ||
+             *_closeTabCommand ||
+             *_closeTabShort ||
+             *_renameTabCommand ||
+             *_renameTabShort ||
              *_focusTabCommand ||
              *_focusTabShort ||
              *_moveFocusCommand ||
@@ -815,6 +821,9 @@ void AppCommandlineArgs::_resetStateToDefault()
     _focusPaneTarget = -1;
     _loadPersistedLayoutIdx = -1;
 
+    _closeTabIndex = -1;
+    _renameTabTitle.clear();
+
     // DON'T clear _launchMode here! This will get called once for every
     // subcommand, so we don't want `wt -F new-tab ; split-pane` clearing out
     // the "global" fullscreen flag (-F).
@@ -1180,3 +1189,66 @@ std::string_view AppCommandlineArgs::GetTargetWindow() const noexcept
 {
     return _windowTarget;
 }
+
+// Method Description:
+// - Adds the `close-tab` subcommand and related options to the commandline parser.
+// - Additionally adds the `ct` subcommand, which is a shortened version of `close-tab`.
+void AppCommandlineArgs::_buildCloseTabParser()
+{
+    _closeTabCommand = _app.add_subcommand("close-tab", RS_A(L"CmdCloseTabDesc"));
+    _closeTabShort = _app.add_subcommand("ct", RS_A(L"CmdCTDesc"));
+
+    auto setupSubcommand = [this](auto* subcommand) {
+        subcommand->add_option("-t,--target",
+                               _closeTabIndex,
+                               RS_A(L"CmdCloseTabTargetArgDesc"));
+
+        subcommand->callback([&, this]() {
+            ActionAndArgs closeTabAction{};
+            closeTabAction.Action(ShortcutAction::CloseTab);
+            if (_closeTabIndex >= 0)
+            {
+                CloseTabArgs args{ static_cast<uint32_t>(_closeTabIndex) };
+                closeTabAction.Args(args);
+            }
+            _startupActions.push_back(closeTabAction);
+        });
+    };
+
+    setupSubcommand(_closeTabCommand);
+    setupSubcommand(_closeTabShort);
+}
+
+// Method Description:
+// - Adds the `rename-tab` subcommand and related options to the commandline parser.
+// - Additionally adds the `rt` subcommand, which is a shortened version of `rename-tab`.
+void AppCommandlineArgs::_buildRenameTabParser()
+{
+    _renameTabCommand = _app.add_subcommand("rename-tab", RS_A(L"CmdRenameTabDesc"));
+    _renameTabShort = _app.add_subcommand("rt", RS_A(L"CmdRTDesc"));
+
+    auto setupSubcommand = [this](auto* subcommand) {
+        subcommand->add_option("title",
+                               _renameTabTitle,
+                               RS_A(L"CmdRenameTabTitleArgDesc"));
+        subcommand->add_option("-t,--target",
+                               _closeTabIndex,
+                               RS_A(L"CmdRenameTabTargetArgDesc"));
+
+        subcommand->callback([&, this]() {
+            ActionAndArgs renameTabAction{};
+            renameTabAction.Action(ShortcutAction::RenameTab);
+            winrt::hstring title{ winrt::to_hstring(_renameTabTitle) };
+            RenameTabArgs args{ title };
+            if (_closeTabIndex >= 0)
+            {
+                args.Index(static_cast<uint32_t>(_closeTabIndex));
+            }
+            renameTabAction.Args(args);
+            _startupActions.push_back(renameTabAction);
+        });
+    };
+
+    setupSubcommand(_renameTabCommand);
+    setupSubcommand(_renameTabShort);
+}
diff --git a/src/cascadia/TerminalApp/AppCommandlineArgs.h b/src/cascadia/TerminalApp/AppCommandlineArgs.h
index 0a9de3999..4b32ba7d6 100644
--- a/src/cascadia/TerminalApp/AppCommandlineArgs.h
+++ b/src/cascadia/TerminalApp/AppCommandlineArgs.h
@@ -93,6 +93,10 @@ private:
     CLI::App* _focusPaneCommand;
     CLI::App* _focusPaneShort;
     CLI::App* _saveCommand;
+    CLI::App* _closeTabCommand;
+    CLI::App* _closeTabShort;
+    CLI::App* _renameTabCommand;
+    CLI::App* _renameTabShort;
 
     // Are you adding a new sub-command? Make sure to update _noCommandsProvided!
 
@@ -125,6 +129,8 @@ private:
     int _focusPaneTarget{ -1 };
     std::string _saveInputName;
     std::string _keyChordOption;
+    int _closeTabIndex{ -1 };
+    std::string _renameTabTitle{};
     // Are you adding more args here? Make sure to reset them in _resetStateToDefault
 
     const Commandline* _currentCommandline{ nullptr };
@@ -150,6 +156,8 @@ private:
     void _buildMovePaneParser();
     void _buildSwapPaneParser();
     void _buildFocusPaneParser();
+    void _buildCloseTabParser();
+    void _buildRenameTabParser();
     bool _noCommandsProvided();
     void _resetStateToDefault();
     int _handleExit(const CLI::App& command, const CLI::Error& e);
diff --git a/src/cascadia/TerminalApp/Resources/en-US/Resources.resw b/src/cascadia/TerminalApp/Resources/en-US/Resources.resw
index 81ca7d51f..2fcf3d486 100644
--- a/src/cascadia/TerminalApp/Resources/en-US/Resources.resw
+++ b/src/cascadia/TerminalApp/Resources/en-US/Resources.resw
@@ -424,6 +424,29 @@
   <data name="CmdSavedLayoutArgDesc" xml:space="preserve">
     <value>This parameter is an internal implementation detail and should not be used.</value>
   </data>
+  <data name="CmdCloseTabDesc" xml:space="preserve">
+    <value>Close a tab</value>
+  </data>
+  <data name="CmdCTDesc" xml:space="preserve">
+    <value>An alias for the "close-tab" subcommand.</value>
+    <comment>{Locked="\"close-tab\""}</comment>
+  </data>
+  <data name="CmdCloseTabTargetArgDesc" xml:space="preserve">
+    <value>Close the tab at the given index (closes the active tab if omitted)</value>
+  </data>
+  <data name="CmdRenameTabDesc" xml:space="preserve">
+    <value>Rename a tab</value>
+  </data>
+  <data name="CmdRTDesc" xml:space="preserve">
+    <value>An alias for the "rename-tab" subcommand.</value>
+    <comment>{Locked="\"rename-tab\""}</comment>
+  </data>
+  <data name="CmdRenameTabTitleArgDesc" xml:space="preserve">
+    <value>The new title for the tab</value>
+  </data>
+  <data name="CmdRenameTabTargetArgDesc" xml:space="preserve">
+    <value>Rename the tab at the given index (renames the active tab if omitted)</value>
+  </data>
   <data name="CmdWindowTargetArgDesc" xml:space="preserve">
     <value>Specify a terminal window to run the given commandline in. "0" always refers to the current window. </value>
   </data>
diff --git a/src/cascadia/TerminalSettingsModel/ActionArgs.h b/src/cascadia/TerminalSettingsModel/ActionArgs.h
index 45ba2cb14..f6145969b 100644
--- a/src/cascadia/TerminalSettingsModel/ActionArgs.h
+++ b/src/cascadia/TerminalSettingsModel/ActionArgs.h
@@ -906,7 +906,11 @@ namespace winrt::Microsoft::Terminal::Settings::Model::implementation
 
     ACTION_ARGS_STRUCT(SetTabColorArgs, SET_TAB_COLOR_ARGS);
 
-    ACTION_ARGS_STRUCT(RenameTabArgs, RENAME_TAB_ARGS);
+    struct RenameTabArgs : public RenameTabArgsT<RenameTabArgs>
+    {
+        ACTION_ARG_BODY(RenameTabArgs, RENAME_TAB_ARGS)
+        ACTION_ARG(Windows::Foundation::IReference<uint32_t>, Index, nullptr);
+    };
 
     ACTION_ARGS_STRUCT(ExecuteCommandlineArgs, EXECUTE_COMMANDLINE_ARGS);
 
diff --git a/src/cascadia/TerminalSettingsModel/ActionArgs.idl b/src/cascadia/TerminalSettingsModel/ActionArgs.idl
index 67c241e23..e81db47e9 100644
--- a/src/cascadia/TerminalSettingsModel/ActionArgs.idl
+++ b/src/cascadia/TerminalSettingsModel/ActionArgs.idl
@@ -312,6 +312,7 @@ namespace Microsoft.Terminal.Settings.Model
     {
         RenameTabArgs(String title);
         String Title { get; };
+        Windows.Foundation.IReference<UInt32> Index { get; };
     };
 
     [default_interface] runtimeclass ExecuteCommandlineArgs : IActionArgs, IActionArgsDescriptorAccess
